File: hydra-haskell/src/main/haskell/Hydra/Impl/Haskell/Dsl/Lib/Math.hs
module Hydra.Impl.Haskell.Dsl.Lib.Math where

import Hydra.Impl.Haskell.Dsl.Phantoms
import qualified Hydra.Impl.Haskell.Dsl.Terms as Terms
import Hydra.Impl.Haskell.Sources.Libraries


add :: Datum (Int -> Int -> Int)
add = Datum $ Terms.primitive _math_add

div :: Datum (Int -> Int -> Int)
div = Datum $ Terms.primitive _math_div

mod :: Datum (Int -> Int -> Int)
mod = Datum $ Terms.primitive _math_mod

mul :: Datum (Int -> Int -> Int)
mul = Datum $ Terms.primitive _math_mul

neg :: Datum (Int -> Int)
neg = Datum $ Terms.primitive _math_neg

rem :: Datum (Int -> Int -> Int)
rem = Datum $ Terms.primitive _math_rem

sub :: Datum (Int -> Int -> Int)
sub = Datum $ Terms.primitive _math_sub

File: hydra-java/src/main/java/hydra/dsl/prims/Math.java
package hydra.dsl.prims;

import hydra.core.Term;
import hydra.lib.math.Add;
import hydra.lib.math.Div;
import hydra.lib.math.Mod;
import hydra.lib.math.Mul;
import hydra.lib.math.Neg;
import hydra.lib.math.Rem;
import hydra.lib.math.Sub;

/**
 * DSL interface providing mathematical primitive operations.
 */
public interface Math {
    /**
     * Returns a term representing the add primitive operation.
     *
     * @return a term for adding two numeric values
     */
    static Term add() {
        return new Add().term();
    }

    /**
     * Returns a term representing the div primitive operation.
     *
     * @return a term for dividing two numeric values
     */
    static Term div() {
        return new Div().term();
    }

    /**
     * Returns a term representing the mod primitive operation.
     *
     * @return a term for computing the modulus of two values
     */
    static Term mod() {
        return new Mod().term();
    }

    /**
     * Returns a term representing the mul primitive operation.
     *
     * @return a term for multiplying two numeric values
     */
    static Term mul() {
        return new Mul().term();
    }

    /**
     * Returns a term representing the neg primitive operation.
     *
     * @return a term for negating a numeric value
     */
    static Term neg() {
        return new Neg().term();
    }

    /**
     * Returns a term representing the rem primitive operation.
     *
     * @return a term for computing the remainder of two values
     */
    static Term rem() {
        return new Rem().term();
    }

    /**
     * Returns a term representing the sub primitive operation.
     *
     * @return a term for subtracting two numeric values
     */
    static Term sub() {
        return new Sub().term();
    }
}
